Implement interpolation of global metrics

Hoolean opened the pull request:
This PR moves the representation of metrics as deltas earlier, to allow us to calculate interpolated metrics.

This has a few use cases; mostly it should help with #1402 and #1489.

Most of the diff is from moving the conversion earlier, and updating tests to use the new builder pattern. Aside from this there are very few changes, aside from a few `(ufo|glyphs)2ir` test constants that needed updating to expect the IR to contain rounded metrics values already.

I anticipate that this PR will need a little more tidying and documentation, and so am submitting it now to get feedback on the general direction first :)

cmyr commented on fontir/src/ir.rs:
```diff
@@ -274,14 +278,17 @@ impl From<KernGroup> for KernSide {
-    pub(crate) HashMap<GlobalMetric, HashMap<NormalizedLocation, OrderedFloat<f64>>>,
-);
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub struct GlobalMetrics(HashMap<GlobalMetric, GlobalMetricValues>);
```
Can we add docs to the effect that "This type is constructed with a [`GlobalMetricsBuilder`]"?

cmyr commented on fontir/src/ir.rs:
```diff
@@ -274,14 +278,17 @@ impl From<KernGroup> for KernSide {
 }
 
+#[derive(Default, Debug)]
+pub struct GlobalMetricsBuilder(
```
In general I encourage docs for all public types.

cmyr commented on fontir/src/ir.rs:
```diff
@@ -553,6 +551,48 @@ impl GlobalMetrics {
         }
     }
 
+    pub fn build(self, axes: &Axes) -> Result<GlobalMetrics, Error> {
```
- I would probably rename `set` above to something more builder-y, like 'add'?
- I would document the `build` method, and in particular I would describe the circumstances in which it will return an error.

Hoolean replied:
For the former: I've now documented the `build` function, and the main error cases to consider.

For the latter: I've left this as `set` to reflect that you can clobber existing values with it, but updated the docs in kind; I tried out the alternative name as normally `add` shouts build to me more too, but along with the previous concern it also made the patch a lot heavier, and so I think we should be fine to keep it as-is if we're happy to.

cmyr reviewed (approved):
Sorry to let this sit, it I think this is a very good idea!

Hoolean commented:
Sorry for the delay, not had much personal time in the past few weeks to push this forward but back and at 'em now :rocket:

I've rebased, updated the documentation with the review in mind, and fixed a few outstanding renames or TODOs that were owed after the refactor. If the tests all pass, then we may be ready to go :)

cmyr commented on fontir/src/ir.rs:
```diff
@@ -284,15 +288,22 @@ impl From<KernGroup> for KernSide {
 }
 
-/// Global metrics. Ascender/descender, cap height, etc.
+/// Used to construct a global metrics variation space [GlobalMetrics], from
```
```suggestion
/// Used to construct a [GlobalMetrics] variation space, from
```

cmyr commented on fontir/src/ir.rs:
```diff
@@ -387,11 +398,12 @@ fn adjust_offset(offset: f64, angle: f64) -> f64 {
-/// Map of values associated with a given global metric at various locations.
-pub type GlobalMetricValues = HashMap<NormalizedLocation, OrderedFloat<f64>>;
+/// Map of deltas associated with a given global metric across regions.
+pub type GlobalMetricDeltas = Vec<(VariationRegion, Vec<f64>)>;
```
this is so slightly confusing... presumably the values at the default location are not deltas? And the values at other locations are relative to the default? This might need to be spelled out more explicitly.

cmyr commented on fontir/src/ir.rs:
```diff
@@ -543,6 +548,8 @@ impl GlobalMetrics {
         self.values_mut(metric).insert(pos, value.into());
     }
 
+    /// Specify the value of a metric at a specific location, but only if there
+    /// is not one already provided. Returns the winning value.
```
Treat docs like git commits: one short summary line, then an empty line, then any elaboration as required, e.g.:

```suggestion
    /// Set the value of a metric at a specific location, if none exists.
    ///
    /// If a value already exists at this location, it is not changed.
    /// 
    /// Returns the (potentially new, potentially preexisting) value.
```

cmyr commented on fontir/src/ir.rs:
```diff
@@ -552,6 +559,9 @@ impl GlobalMetrics {
         *self.values_mut(metric).entry(pos).or_insert(value.into())
     }
 
+    /// Specify the value of a metric at a specific location, if we have one;
+    /// this is a helper build method to reduce boilerplate. If there is an
+    /// existing value, it will be overwritten.
```
```suggestion
    /// A helper method for setting possibly missing values.
    ///
    ///
    /// This is a helper build method to reduce boilerplate. If there is an
    /// existing value, it will be overwritten.
```

Hoolean replied:
Trimmed extra line locally and applied

cmyr commented on fontir/src/ir.rs:
```diff
@@ -563,6 +573,60 @@ impl GlobalMetrics {
+}
+
+impl GlobalMetrics {
+    /// Access the raw deltas for a particular metric. This lower level of
+    /// access is useful for interpolating the value at an invididual positition
+    /// manually, and testing.
```
```suggestion
    /// Access the raw deltas for a particular metric.
    ///
    /// This lower level of access is useful for interpolating the value at an
    /// invididual positition manually, and testing.
```

cmyr commented on fontir/src/ir.rs:
```diff
@@ -602,7 +666,9 @@ impl GlobalMetrics {
     }
 
-    pub fn iter(&self) -> impl Iterator<Item = (&GlobalMetric, &GlobalMetricValues)> + '_ {
+    /// Access the raw deltas for every metric. This lower level of access is
+    /// most useful for serialising the variation spaces.
```
```suggestion
    /// Access the raw deltas for every metric.
    ///
    /// This lower level of access is most useful for serialising the variation spaces.
```

Hoolean replied:
Applied locally

cmyr reviewed (approved):
Looks great! Some little nits about docs inline, but happy to merge this :)

Hoolean commented:
Addressed feedback – I've left the suggestions unresolved in case you want to double-check any
